src/SAWScript/SolverCache.hs

```diff
@@ -0,0 +1,483 @@
+    vs      <- v .:  "vs"
+    opts    <- v .:? "opts"
+    nm      <- v .:  "nm"
+    mb_cexs <- v .:? "cexs"
+    return $ SolverCacheValue vs (fromMaybe [] opts) nm mb_cexs
+
+instance ToJSON SolverCacheValue where
+  toJSON (SolverCacheValue vs opts nm mb_cexs) = JSON.object $
+    ["vs" .= vs] ++ (if null opts then [] else ["opts" .= opts]) ++
+    ["nm" .= nm] ++ maybe [] (\cexs -> ["cexs" .= cexs]) mb_cexs
+  toEncoding (SolverCacheValue vs opts nm mb_cexs) = JSON.pairs $
+    "vs" .= vs <> (if null opts then mempty else "opts" .= opts) <>
+    "nm" .= nm <> maybe mempty (\cexs -> "cexs" .= cexs) mb_cexs
+
+-- | Convert the result of a solver call on the given 'SATQuery' to a
+-- 'SolverCacheValue'
+toSolverCacheValue :: SolverBackendVersions -> [SolverBackendOption] ->
+                      SATQuery -> (Maybe CEX, String) -> Maybe SolverCacheValue
+toSolverCacheValue vs opts satq (cexs, solver_name) =
+  fmap (SolverCacheValue vs opts solver_name)
+       (mapM (mapM (firstM (`elemIndex` ecs))) cexs)
+  where ecs = M.keys $ satVariables satq
+
+-- | Convert a 'SolverCacheValue' to something which has the same form as the
+-- result of a solver call on the given 'SATQuery'
+fromSolverCacheValue :: SATQuery -> SolverCacheValue -> (Maybe CEX, String)
+fromSolverCacheValue satq (SolverCacheValue _ _ solver_name cexs) =
+  (fmap (fmap (first (ecs !!))) cexs, solver_name)
+  where ecs = M.keys $ satVariables satq
+
+
+-- The Solver Cache ------------------------------------------------------------
+
+-- | A 'SolverCacheDB' of cached solver results as well as counters for how
+-- many cache hits and how many new entry creations have occurred
+data SolverCache =
+  SolverCache
+  { solverCacheDB      :: LMDBOptDatabase SolverCacheValue
+  , solverCacheStats   :: IORef (Map SolverCacheStat Integer)
+  , solverCacheTimeout :: Int
+  }
+
+-- | A statistic to track in 'solverCacheStats'
+data SolverCacheStat = Lookups | FailedLookups | Inserts | FailedInserts
+                     deriving (Eq, Ord, Bounded, Enum)
+
+-- | An empty 'SolverCache' with no associated 'FilePath'
+emptySolverCache :: IO SolverCache
+emptySolverCache = do
+  db <- LMDBOpt.new
+  stats <- newIORef $ M.fromList ((,0) <$> [minBound..])
+  return $ SolverCache db stats 1000
+
+-- | An operation on a 'SolverCache', returning a value of the given type or
+-- returning an optional default value in the case of no enabled 'SolverCache'
+type SolverCacheOp a = (Maybe a, Options -> SolverCache -> IO a)
+
+-- | Lookup a 'SolverCacheKey' in the solver result cache
+lookupInSolverCache :: SolverCacheKey -> SolverCacheOp (Maybe SolverCacheValue)
+lookupInSolverCache k = (Just Nothing,) $ \opts SolverCache{..} ->
+  tryWithTimeout solverCacheTimeout
+                 (LMDBOpt.lookup (solverCacheKeyHash k)
+                                 solverCacheDB) >>= \case
+    Right (Just v) -> do
+      printOutLn opts Info ("Using cached result: " ++ show k)
```

Comment:
This output can be useful, but I wonder if it is too noisy for most users. Is it worth guarding this behind a higher `-v` level?

Reply:
Good idea! I added them to the `Debug` verbosity level.